File: examples/08_metaprogramming_patterns.ts
function main(): void {

  const FIELD_COUNT: number = 3;

  const TYPE_NAME: string = "Point3D";

interface Point3D {
  x: number;
  y: number;
  z: number;
}

  function createPoint3D(x: number, y: number, z: number): Point3D {
    return {
      x: x,
      y: y,
      z: z
    };
  }

  const POINT3_D_SIZE: number = 3;

  console.log(`${Point3D.type_name()} has ${Point3D.field_count()} fields`);

  const VARIANT_A: number = 1;

  const VARIANT_B: number = 2;

  enum Tag {
    A,
    B,
  }
  let tag = Tag.A;
  console.log(`tag discriminant: ${tag}`);
}

main();

File: examples/08_metaprogramming_patterns.rs
pub fn main() -> () {
    pub const FIELD_COUNT: usize = 3;
    pub const TYPE_NAME: &str = "Point3D".to_string();
    pub struct Point3D {
        pub x: i64,
        pub y: i64,
        pub z: i64,
    }
    impl Point3D {
        pub fn type_name() -> &str {
            TYPE_NAME
        }
        pub fn field_count() -> usize {
            FIELD_COUNT
        }
    }
    println!(
        "{} has {} fields",
        Point3D::type_name(),
        Point3D::field_count()
    );
    pub const VARIANT_A: u8 = 1;
    pub const VARIANT_B: u8 = 2;
    pub enum Tag {
        A = VARIANT_A as isize,
        B = VARIANT_B as isize,
    }
    let tag = Tag::A;
    println!("tag discriminant: {}", tag as u8);
}
